Hello @shahor02 , 
The o2 CI fails, but I don't know why, maybe some incompatibility with QC?
```
[3336:TST-MNLTest53834-proxy]: [INFO] Instrumenting crash signals
[3336:TST-MNLTest53834-proxy]: [INFO] Spawing new device TST-MNLTest53834-proxy in process with pid 3336
[3336:TST-MNLTest53834-proxy]: [16:12:01][INFO] 
[3336:TST-MNLTest53834-proxy]:       ______      _    _______  _________ 
[3336:TST-MNLTest53834-proxy]:      / ____/___ _(_)_______   |/  /_  __ \    version 1.8.2
[3336:TST-MNLTest53834-proxy]:     / /_  / __ `/ / ___/__  /|_/ /_  / / /    build RELWITHDEBINFO
[3336:TST-MNLTest53834-proxy]:    / __/ / /_/ / / /    _  /  / / / /_/ /     https://github.com/FairRootGroup/FairMQ
[3336:TST-MNLTest53834-proxy]:   /_/    \__,_/_/_/     /_/  /_/  \___\_\     LGPL-3.0  © 2012-2023 GSI
[3336:TST-MNLTest53834-proxy]: 
[3336:TST-MNLTest53834-proxy]: [16:12:01][STATE] Starting FairMQ state machine --> IDLE
[3336:TST-MNLTest53834-proxy]: [16:12:01][STATE] IDLE ---> INITIALIZING DEVICE
[3336:TST-MNLTest53834-proxy]: [16:12:01][INFO] Drain mode requested while in Ready state
[3336:TST-MNLTest53834-proxy]: [16:12:01][STATE] INITIALIZING DEVICE ---> INITIALIZED
[3336:TST-MNLTest53834-proxy]: [16:12:01][STATE] INITIALIZED ---> BINDING
[3336:TST-MNLTest53834-proxy]: [16:12:01][ERROR] failed to attach channel TST-MNLTest53834-proxy[0] (bind on tcp://*:53834)
[3336:TST-MNLTest53834-proxy]: [16:12:01][ERROR] 1 of the binding channels could not initialize. Initial configuration incomplete.
[3336:TST-MNLTest53834-proxy]: [16:12:01][STATE] Exiting FairMQ state machine
[3336:TST-MNLTest53834-proxy]: [16:12:01][FATAL] error while setting up workflow in o2-qc: 1 of the binding channels could not initialize. Initial configuration incomplete.
[3336:TST-MNLTest53834-proxy]: terminate called after throwing an instance of 'o2::framework::RuntimeErrorRef'
[3336:TST-MNLTest53834-proxy]: *** Program crashed (Segmentation fault, FPE, BUS, ABRT, KILL, Unhandled Exception, ...)
[3336:TST-MNLTest53834-proxy]: Backtrace by DPL:
[3336:TST-MNLTest53834-proxy]: ABRT
[3336:TST-MNLTest53834-proxy]: Executable is /sw/INSTALLROOT/3dda32aba7f7da12d77a586052c949f3d9535f28/slc7_x86-64/QualityControl/v1.122.0-local28/bin/o2-qc
[3336:TST-MNLTest53834-proxy]:     /lib64/libc.so.6:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /lib64/libc.so.6:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /lib64/libc.so.6:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/GCC-Toolchain/v12.2.0-alice1-9/lib64/libstdc++.so.6(+0xa58d9) [0x7efca48e88d9]:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/GCC-Toolchain/v12.2.0-alice1-9/lib64/libstdc++.so.6(+0xb110a) [0x7efca48f410a]:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/GCC-Toolchain/v12.2.0-alice1-9/lib64/libstdc++.so.6(+0xb1175) [0x7efca48f4175]:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/GCC-Toolchain/v12.2.0-alice1-9/lib64/libstdc++.so.6(+0xb13c8) [0x7efca48f43c8]:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/O2/12278-slc7_x86-64-local1/lib/libO2Framework.so:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/FairLogger/v1.11.1-15/lib/libFairLogger.so.1.11: fair::Logger::~Logger()
[3336:TST-MNLTest53834-proxy]:     /sw/slc7_x86-64/O2/12278-slc7_x86-64-local1/lib/libO2Framework.so: doUnknownException(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*)
[3336:TST-MNLTest53834-proxy]:     o2-qc() [0x4169a0]:     std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() at basic_string.h:795
[3336:TST-MNLTest53834-proxy]:     /lib64/libc.so.6:     ?? ??:0
[3336:TST-MNLTest53834-proxy]:     o2-qc() [0x416fb4]:     _start at ??:?
[3336:TST-MNLTest53834-proxy]: Backtrace complete.
[ERROR] pid 3336 (TST-MNLTest53834-proxy) crashed with or was killed with exit code 1
[INFO] QUIT_REQUESTED
```
@knopers8 , @Barthelemy , @ktf , any idea?

--- a/Detectors/GlobalTracking/include/GlobalTracking/MatchTOF.h
+++ b/Detectors/GlobalTracking/include/GlobalTracking/MatchTOF.h
@@ -201,10 +201,14 @@ class MatchTOF
 
   void storeMatchable(bool val = true) { mStoreMatchable = val; }
 
+  void setNlanes(int lanes) { mNlanes = lanes; }
+
  private:
   bool prepareFITData();
   int prepareInteractionTimes();
   bool prepareTPCData();
+  void propagateTPCTracks(int sec);
+  void propagateConstrTracks(int sec);
   void addTPCSeed(const o2::tpc::TrackTPC& _tr, o2::dataformats::GlobalTrackID srcGID, float time0, float terr);
   void addITSTPCSeed(const o2::dataformats::TrackTPCITS& _tr, o2::dataformats::GlobalTrackID srcGID, float time0, float terr);
   void addTRDSeed(const o2::trd::TrackTRD& _tr, o2::dataformats::GlobalTrackID srcGID, float time0, float terr);
@@ -215,7 +219,7 @@ class MatchTOF
 
   void doMatching(int sec);
   void doMatchingForTPC(int sec);
-  void selectBestMatches();
+  void selectBestMatches(int sec);
   void BestMatches(std::vector<o2::dataformats::MatchInfoTOFReco>& matchedTracksPairs, std::vector<o2::dataformats::MatchInfoTOF>* matchedTracks, std::vector<int>* matchedTracksIndex, int* matchedClustersIndex, const gsl::span<const o2::ft0::RecPoints>& FITRecPoints, const std::vector<Cluster>& TOFClusWork, const std::vector<matchTrack>* TracksWork, std::vector<o2::dataformats::CalibInfoTOF>& CalibInfoTOF, unsigned long Timestamp, bool MCTruthON, const o2::dataformats::MCTruthContainer<o2::MCCompLabel>* TOFClusLabels, const std::vector<o2::MCCompLabel>* TracksLblWork, std::vector<o2::MCCompLabel>* OutTOFLabels, float calibMaxChi2);
   void BestMatchesHP(std::vector<o2::dataformats::MatchInfoTOFReco>& matchedTracksPairs, std::vector<o2::dataformats::MatchInfoTOF>* matchedTracks, std::vector<int>* matchedTracksIndex, int* matchedClustersIndex, const gsl::span<const o2::ft0::RecPoints>& FITRecPoints, const std::vector<Cluster>& TOFClusWork, std::vector<o2::dataformats::CalibInfoTOF>& CalibInfoTOF, unsigned long Timestamp, bool MCTruthON, const o2::dataformats::MCTruthContainer<o2::MCCompLabel>* TOFClusLabels, const std::vector<o2::MCCompLabel>* TracksLblWork, std::vector<o2::MCCompLabel>* OutTOFLabels);
   bool propagateToRefX(o2::track::TrackParCov& trc, float xRef /*in cm*/, float stepInCm /*in cm*/, o2::track::TrackLTIntegral& intLT);
@@ -231,6 +235,8 @@ class MatchTOF
   // Data members
   const o2::globaltracking::RecoContainer* mRecoCont = nullptr;
   o2::InteractionRecord mStartIR{0, 0}; ///< IR corresponding to the start of the TF
+
+  int mNlanes = 3; ///< for multi-threading in matching
 
   // TOF matching params (work in progress)
   const MatchTOFParams* mMatchParams = nullptr;
@@ -299,22 +305,22 @@ class MatchTOF
   /// <<<-----
 
   ///<working copy of the input tracks
-  std::vector<matchTrack> mTracksWork[trkType::SIZE]; ///<track params prepared for matching + time value
-  std::vector<o2::MCCompLabel> mTracksLblWork[trkType::SIZE];           ///<TPCITS track labels
-  std::vector<o2::track::TrackLTIntegral> mLTinfos[trkType::SIZE];      ///<expected times and others
-  std::vector<o2::dataformats::GlobalTrackID> mTrackGid[trkType::SIZE]; ///<expected times and others
+  std::vector<matchTrack> mTracksWork[o2::constants::math::NSectors][trkType::SIZE];                   ///< track params prepared for matching + time value
+  std::vector<o2::MCCompLabel> mTracksLblWork[o2::constants::math::NSectors][trkType::SIZE];           ///< TPCITS track labels
+  std::vector<o2::track::TrackLTIntegral> mLTinfos[o2::constants::math::NSectors][trkType::SIZE];      ///< expected times and others
+  std::vector<o2::dataformats::GlobalTrackID> mTrackGid[o2::constants::math::NSectors][trkType::SIZE]; ///< expected times and others
   ///< per sector indices of track entry in mTracksWork
   std::array<std::vector<int>, o2::constants::math::NSectors> mTracksSectIndexCache[trkType::SIZE];
+  std::array<std::vector<int>, o2::constants::math::NSectors> mTracksSeed[trkType::SIZE];
 
-  std::vector<float> mExtraTPCFwdTime;                             ///<track extra params for TPC tracks: Fws Max time
-  std::vector<Cluster> mTOFClusWork;                               ///<track params prepared for matching
-  std::vector<int8_t> mSideTPC;                                    ///<track side for TPC tracks
+  std::vector<float> mExtraTPCFwdTime[o2::constants::math::NSectors]; ///< track extra params for TPC tracks: Fws Max time
+  std::vector<Cluster> mTOFClusWork;                                  ///< track params prepared for matching
+  std::vector<int8_t> mSideTPC[o2::constants::math::NSectors];        ///< track side for TPC tracks
 
   ///< per sector indices of TOF cluster entry in mTOFClusWork
   std::array<std::vector<int>, o2::constants::math::NSectors> mTOFClusSectIndexCache;
 
-  ///<array of track-TOFCluster pairs from the matching
-  std::vector<o2::dataformats::MatchInfoTOFReco> mMatchedTracksPairs;
+  ///< array of track-TOFCluster pairs from the matching
   std::vector<o2::dataformats::MatchInfoTOFReco> mMatchedTracksPairsSec[o2::constants::math::NSectors];
 
   ///<array of TOFChannel calibration info
@@ -325,7 +331,7 @@ class MatchTOF
   std::vector<o2::dataformats::MatchInfoTOF> mMatchedTracks[trkType::SIZEALL]; // this is the output of the matching -> UNCONS, CONSTR
   std::vector<o2::MCCompLabel> mOutTOFLabels[trkType::SIZEALL];                ///< TOF label of matched tracks
 
-  std::vector<int> mMatchedTracksIndex[trkType::SIZE]; // vector of indexes of the tracks to be matched
+  std::vector<int> mMatchedTracksIndex[o2::constants::math::NSectors][trkType::SIZE]; // vector of indexes of the tracks to be matched
 
   int mNumOfClusters;                   // number of clusters to be matched
   int* mMatchedClustersIndex = nullptr; //[mNumOfClusters]

--- a/Detectors/GlobalTracking/src/MatchTOF.cxx
+++ b/Detectors/GlobalTracking/src/MatchTOF.cxx
@@ -46,6 +46,10 @@
 #include "DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h"
 #include "TOFBase/Utils.h"
 
+#ifdef WITH_OPENMP
+#include <omp.h>
+#endif
+
 using namespace o2::globaltracking;
 using evGIdx = o2::dataformats::EvIndex<int, o2::dataformats::GlobalTrackID>;
 using trkType = o2::dataformats::MatchInfoTOFReco::TrackType;
@@ -84,23 +88,24 @@ void MatchTOF::run(const o2::globaltracking::RecoContainer& inp, unsigned long f
     mOutTOFLabels[i].clear();
   }
 
-  for (int i = 0; i < trkType::SIZE; i++) {
-    mTracksWork[i].clear();
-    mTrackGid[i].clear();
-  }
   for (int it = 0; it < trkType::SIZE; it++) {
-    mMatchedTracksIndex[it].clear();
-    mLTinfos[it].clear();
-    if (mMCTruthON) {
-      mTracksLblWork[it].clear();
-    }
-    for (int sec = o2::constants::math::NSectors; sec--;) {
+    for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
+      mMatchedTracksIndex[sec][it].clear();
+      mTracksWork[sec][it].clear();
+      mTrackGid[sec][it].clear();
+      mLTinfos[sec][it].clear();
+      if (mMCTruthON) {
+        mTracksLblWork[sec][it].clear();
+      }
       mTracksSectIndexCache[it][sec].clear();
+      mTracksSeed[it][sec].clear();
     }
   }
 
-  mSideTPC.clear();
-  mExtraTPCFwdTime.clear();
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
+    mSideTPC[sec].clear();
+    mExtraTPCFwdTime[sec].clear();
+  }
 
   mTimerTot.Start();
   bool isPrepareTOFClusters = prepareTOFClusters();
@@ -127,32 +132,46 @@ void MatchTOF::run(const o2::globaltracking::RecoContainer& inp, unsigned long f
 
   mTimerTot.Start();
   std::array<uint32_t, 18> nMatches = {0};
-  for (int sec = o2::constants::math::NSectors; sec--;) {
-    mMatchedTracksPairs.clear(); // new sector
-    LOG(debug) << "Doing matching for sector " << sec << "...";
-    if (mIsITSTPCused || mIsTPCTRDused || mIsITSTPCTRDused) {
-      mTimerMatchITSTPC.Start(sec == o2::constants::math::NSectors - 1);
+  // run matching (this can be multi-thread)
+  mTimerMatchITSTPC.Start();
+
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
+    mMatchedTracksPairsSec[sec].clear(); // new sector
+  }
+
+  o2::tof::Geo::Init();
+
+  if (mIsITSTPCused || mIsTPCTRDused || mIsITSTPCTRDused) {
+#ifdef WITH_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(mNlanes)
+#endif
+    for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
       doMatching(sec);
-      mTimerMatchITSTPC.Stop();
     }
-    if (mIsTPCused) {
-      mTimerMatchTPC.Start(sec == o2::constants::math::NSectors - 1);
+  }
+  mTimerMatchITSTPC.Stop();
+
+  mTimerMatchTPC.Start();
+  if (mIsTPCused) {
+#ifdef WITH_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(mNlanes)
+#endif
+    for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
       doMatchingForTPC(sec);
-      mTimerMatchTPC.Stop();
     }
+  }
+  mTimerMatchTPC.Stop();
 
+  // finalize
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
     if (mStoreMatchable) {
-      // fill per sector
-      mMatchedTracksPairsSec[sec].clear();
-      mMatchedTracksPairsSec[sec].insert(mMatchedTracksPairsSec[sec].end(), mMatchedTracksPairs.begin(), mMatchedTracksPairs.end());
-
       // if MC check if good or fake matches
       if (mMCTruthON) {
         for (auto& matchingPair : mMatchedTracksPairsSec[sec]) {
           int trkType = (int)matchingPair.getTrackType();
           int itrk = matchingPair.getIdLocal();
           const auto& labelsTOF = mTOFClusLabels->getLabels(matchingPair.getTOFClIndex());
-          const auto& labelTrack = mTracksLblWork[trkType][itrk];
+          const auto& labelTrack = mTracksLblWork[sec][trkType][itrk];
 
           // we have not found the track label among those associated to the TOF cluster --> fake match! We will associate the label of the main channel, but negative
           bool fake = true;
@@ -169,11 +188,11 @@ void MatchTOF::run(const o2::globaltracking::RecoContainer& inp, unsigned long f
     }
 
     LOG(debug) << "...done. Now check the best matches";
-    nMatches[sec] = mMatchedTracksPairs.size();
-    selectBestMatches();
+    nMatches[sec] = mMatchedTracksPairsSec[sec].size();
+    selectBestMatches(sec);
   }
   std::string nMatchesStr = "Number of pairs matched per sector: ";
-  for (int sec = o2::constants::math::NSectors; sec--;) {
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
     nMatchesStr += fmt::format("{} : {} ; ", sec, nMatches[sec]);
   }
   LOG(info) << nMatchesStr;
@@ -361,24 +380,85 @@ bool MatchTOF::prepareTPCData()
   };
   mRecoCont->createTracksVariadic(creator);
 
+#ifdef WITH_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(mNlanes)
+#endif
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
+    if (mIsTPCused) {
+      propagateTPCTracks(sec);
+    }
+    if (mIsITSTPCused || mIsTPCTRDused || mIsITSTPCTRDused) {
+      propagateConstrTracks(sec);
+    }
+  }
+
+  // re-assign tracks which change sector (no multi-threading)
+  std::array<float, 3> globalPos;
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
+    if (mIsTPCused) {
+      for (auto& it : mTracksSeed[trkType::UNCONS][sec]) {
+        auto& pair = mTracksWork[sec][trkType::UNCONS][it];
+        auto& trc = pair.first;
+        trc.getXYZGlo(globalPos);
+        int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
+
+        int itnew = mTracksWork[sector][trkType::UNCONS].size();
+
+        mSideTPC[sector].push_back(mSideTPC[sec][it]);
+        mExtraTPCFwdTime[sector].push_back(mExtraTPCFwdTime[sec][it]);
+        mTracksWork[sector][trkType::UNCONS].emplace_back(pair);
+        mTrackGid[sector][trkType::UNCONS].emplace_back(mTrackGid[sec][trkType::UNCONS][it]);
+
+        if (mMCTruthON) {
+          mTracksLblWork[sector][trkType::UNCONS].emplace_back(mTracksLblWork[sec][trkType::UNCONS][it]);
+        }
+        mLTinfos[sector][trkType::UNCONS].emplace_back(mLTinfos[sec][trkType::UNCONS][it]);
+        mTracksSectIndexCache[trkType::UNCONS][sector].push_back(itnew);
+      }
+    }
+
+    for (auto& it : mTracksSeed[trkType::CONSTR][sec]) {
+      auto& pair = mTracksWork[sec][trkType::CONSTR][it];
+      auto& trc = pair.first;
+      trc.getXYZGlo(globalPos);
+      int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
+
+      int itnew = mTracksWork[sector][trkType::CONSTR].size();
+
+      mTracksWork[sector][trkType::CONSTR].emplace_back(pair);
+      mTrackGid[sector][trkType::CONSTR].emplace_back(mTrackGid[sec][trkType::CONSTR][it]);
+
+      if (mMCTruthON) {
+        mTracksLblWork[sector][trkType::CONSTR].emplace_back(mTracksLblWork[sec][trkType::CONSTR][it]);
+      }
+      mLTinfos[sector][trkType::CONSTR].emplace_back(mLTinfos[sec][trkType::CONSTR][it]);
+      mTracksSectIndexCache[trkType::CONSTR][sector].push_back(itnew);
+    }
+  }
+
   for (int it = 0; it < trkType::SIZE; it++) {
-    mMatchedTracksIndex[it].resize(mTracksWork[it].size());
-    std::fill(mMatchedTracksIndex[it].begin(), mMatchedTracksIndex[it].end(), -1); // initializing all to -1
+    for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
+      mMatchedTracksIndex[sec][it].resize(mTracksWork[sec][it].size());
+      std::fill(mMatchedTracksIndex[sec][it].begin(), mMatchedTracksIndex[sec][it].end(), -1); // initializing all to -1
+    }
   }
 
   if (mIsTPCused) {
     LOG(debug) << "Number of UNCONSTRAINED tracks that failed to be propagated to TOF = " << mNotPropagatedToTOF[trkType::UNCONS];
 
     // sort tracks in each sector according to their time (increasing in time)
-    for (int sec = o2::constants::math::NSectors; sec--;) {
+#ifdef WITH_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(mNlanes)
+#endif
+    for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
       auto& indexCache = mTracksSectIndexCache[trkType::UNCONS][sec];
       LOG(debug) << "Sorting sector" << sec << " | " << indexCache.size() << " tracks";
       if (!indexCache.size()) {
         continue;
       }
-      std::sort(indexCache.begin(), indexCache.end(), [this](int a, int b) {
-        auto& trcA = mTracksWork[trkType::UNCONS][a].second;
-        auto& trcB = mTracksWork[trkType::UNCONS][b].second;
+      std::sort(indexCache.begin(), indexCache.end(), [this, sec](int a, int b) {
+        auto& trcA = mTracksWork[sec][trkType::UNCONS][a].second;
+        auto& trcB = mTracksWork[sec][trkType::UNCONS][b].second;
         return ((trcA.getTimeStamp() - trcA.getTimeStampError()) - (trcB.getTimeStamp() - trcB.getTimeStampError()) < 0.);
       });
     } // loop over tracks of single sector
@@ -387,21 +467,112 @@ bool MatchTOF::prepareTPCData()
     LOG(debug) << "Number of CONSTRAINED tracks that failed to be propagated to TOF = " << mNotPropagatedToTOF[trkType::CONSTR];
 
     // sort tracks in each sector according to their time (increasing in time)
-    for (int sec = o2::constants::math::NSectors; sec--;) {
+#ifdef WITH_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(mNlanes)
+#endif
+    for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
       auto& indexCache = mTracksSectIndexCache[trkType::CONSTR][sec];
       LOG(debug) << "Sorting sector" << sec << " | " << indexCache.size() << " tracks";
       if (!indexCache.size()) {
         continue;
       }
-      std::sort(indexCache.begin(), indexCache.end(), [this](int a, int b) {
-        auto& trcA = mTracksWork[trkType::CONSTR][a].second;
-        auto& trcB = mTracksWork[trkType::CONSTR][b].second;
+      std::sort(indexCache.begin(), indexCache.end(), [this, sec](int a, int b) {
+        auto& trcA = mTracksWork[sec][trkType::CONSTR][a].second;
+        auto& trcB = mTracksWork[sec][trkType::CONSTR][b].second;
         return ((trcA.getTimeStamp() - mSigmaTimeCut * trcA.getTimeStampError()) - (trcB.getTimeStamp() - mSigmaTimeCut * trcB.getTimeStampError()) < 0.);
       });
     } // loop over tracks of single sector
   }
 
   return true;
+}
+
+//______________________________________________
+void MatchTOF::propagateTPCTracks(int sec)
+{
+  auto& trkWork = mTracksWork[sec][trkType::UNCONS];
+
+  for (int it = 0; it < trkWork.size(); it++) {
+    o2::track::TrackParCov& trc = trkWork[it].first;
+
+    o2::track::TrackLTIntegral& intLT0 = mLTinfos[sec][trkType::UNCONS][it];
+
+    const auto& trackTune = o2::globaltracking::TrackTuneParams::Instance();
+    if (!trackTune.sourceLevelTPC) { // correct only if TPC track was not corrected at the source level
+      if (trackTune.useTPCOuterCorr) {
+        trc.updateParams(trackTune.tpcParOuter);
+      }
+      if (trackTune.tpcCovOuterType != o2::globaltracking::TrackTuneParams::AddCovType::Disable) { // only TRD-refitted track have cov.matrix already man>
+        trc.updateCov(trackTune.tpcCovOuter, trackTune.tpcCovOuterType == o2::globaltracking::TrackTuneParams::AddCovType::WithCorrelations);
+      }
+    }
+    if (!propagateToRefXWithoutCov(trc, mXRef, 10, mBz)) { // we first propagate to 371 cm without considering the covariance matri
+      mNotPropagatedToTOF[trkType::UNCONS]++;
+      continue;
+    }
+
+    if (trc.getX() < o2::constants::geom::XTPCOuterRef - 1.) {
+      if (!propagateToRefX(trc, o2::constants::geom::XTPCOuterRef, 10, intLT0) || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagat>
+        mNotPropagatedToTOF[trkType::UNCONS]++;
+        continue;
+      }
+    }
+
+    o2::base::Propagator::Instance()->estimateLTFast(intLT0, trc);
+
+    // the "rough" propagation worked; now we can propagate considering also the cov matrix
+    if (!propagateToRefX(trc, mXRef, 2, intLT0)) { // || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix w>
+      mNotPropagatedToTOF[trkType::UNCONS]++;
+      continue;
+    }
+
+    //  printf("time0 %f -> %f (diff = %f, err = %f)\n",trackTime0, time0, trackTime0 - time0, terr);
+    //  printf("time errors %f,%f -> %f,%f\n",(_tr.getDeltaTBwd() + 5) * mTPCTBinMUS,(_tr.getDeltaTFwd() + 5) * mTPCTBinMUS,trackTime0 - time0 + terr, ti>
+
+    std::array<float, 3> globalPos;
+    trc.getXYZGlo(globalPos);
+    int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
+
+    if (sector == sec) {
+      mTracksSectIndexCache[trkType::UNCONS][sector].push_back(it);
+    } else {
+      mTracksSeed[trkType::UNCONS][sec].push_back(it); // to be moved to another sector
+    }
+  }
+}
+//______________________________________________
+void MatchTOF::propagateConstrTracks(int sec)
+{
+  std::array<float, 3> globalPos;
+
+  auto& trkWork = mTracksWork[sec][trkType::CONSTR];
+
+  for (int it = 0; it < trkWork.size(); it++) {
+    o2::track::TrackParCov& trc = trkWork[it].first;
+    o2::track::TrackLTIntegral& intLT0 = mLTinfos[sec][trkType::CONSTR][it];
+
+    // propagate to matching Xref
+    if (!propagateToRefXWithoutCov(trc, mXRef, 2, mBz)) { // we first propagate to 371 cm without considering the covariance matrix
+      mNotPropagatedToTOF[trkType::CONSTR]++;
+      continue;
+    }
+
+    // the "rough" propagation worked; now we can propagate considering also the cov matrix
+    if (!propagateToRefX(trc, mXRef, 2, intLT0) || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix worked;>
+      mNotPropagatedToTOF[trkType::CONSTR]++;
+      continue;
+    }
+
+    trc.getXYZGlo(globalPos);
+
+    int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
+
+    if (sector == sec) {
+      mTracksSectIndexCache[trkType::CONSTR][sector].push_back(it);
+    } else {
+      mTracksSeed[trkType::CONSTR][sec].push_back(it); // to be moved to another sector
+    }
+  }
 }
 //______________________________________________
 void MatchTOF::addITSTPCSeed(const o2::dataformats::TrackTPCITS& _tr, o2::dataformats::GlobalTrackID srcGID, float time0, float terr)
@@ -418,44 +589,19 @@ void MatchTOF::addITSTPCSeed(const o2::dataformats::TrackTPCITS& _tr, o2::datafo
 void MatchTOF::addConstrainedSeed(o2::track::TrackParCov& trc, o2::dataformats::GlobalTrackID srcGID, o2::track::TrackLTIntegral intLT0, timeEst timeMUS)
 {
   std::array<float, 3> globalPos;
-  // current track index
-  int it = mTracksWork[trkType::CONSTR].size();
-
-  // propagate to matching Xref
   trc.getXYZGlo(globalPos);
-  LOG(debug) << "Global coordinates Before propagating to 371 cm: globalPos[0] = " << globalPos[0] << ", globalPos[1] = " << globalPos[1] << ", globalPos[2] = " << globalPos[2];
-  LOG(debug) << "Radius xy Before propagating to 371 cm = " << TMath::Sqrt(globalPos[0] * globalPos[0] + globalPos[1] * globalPos[1]);
-  LOG(debug) << "Radius xyz Before propagating to 371 cm = " << TMath::Sqrt(globalPos[0] * globalPos[0] + globalPos[1] * globalPos[1] + globalPos[2] * globalPos[2]);
-  if (!propagateToRefXWithoutCov(trc, mXRef, 2, mBz)) { // we first propagate to 371 cm without considering the covariance matrix
-    mNotPropagatedToTOF[trkType::CONSTR]++;
-    return;
-  }
-
-  // the "rough" propagation worked; now we can propagate considering also the cov matrix
-  if (!propagateToRefX(trc, mXRef, 2, intLT0) || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix worked; CHECK: can it happen that it does not if the prop>
-    mNotPropagatedToTOF[trkType::CONSTR]++;
-    return;
-  }
-
-  trc.getXYZGlo(globalPos);
-
-  LOG(debug) << "Global coordinates After propagating to 371 cm: globalPos[0] = " << globalPos[0] << ", globalPos[1] = " << globalPos[1] << ", globalPos[2] = " << globalPos[2];
-  LOG(debug) << "Radius xy After propagating to 371 cm = " << TMath::Sqrt(globalPos[0] * globalPos[0] + globalPos[1] * globalPos[1]);
-  LOG(debug) << "Radius xyz After propagating to 371 cm = " << TMath::Sqrt(globalPos[0] * globalPos[0] + globalPos[1] * globalPos[1] + globalPos[2] * globalPos[2]);
-  LOG(debug) << "The track will go to sector " << o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
 
   int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
 
   // create working copy of track param
-  mTracksWork[trkType::CONSTR].emplace_back(std::make_pair(trc, timeMUS));
-  mTrackGid[trkType::CONSTR].emplace_back(srcGID);
-  mLTinfos[trkType::CONSTR].emplace_back(intLT0);
+  mTracksWork[sector][trkType::CONSTR].emplace_back(std::make_pair(trc, timeMUS));
+  mTrackGid[sector][trkType::CONSTR].emplace_back(srcGID);
+  mLTinfos[sector][trkType::CONSTR].emplace_back(intLT0);
 
   if (mMCTruthON) {
-    mTracksLblWork[trkType::CONSTR].emplace_back(mRecoCont->getTPCITSTrackMCLabel(srcGID));
+    mTracksLblWork[sector][trkType::CONSTR].emplace_back(mRecoCont->getTPCITSTrackMCLabel(srcGID));
   }
 
-  mTracksSectIndexCache[trkType::CONSTR][sector].push_back(it);
   //delete trc; // Check: is this needed?
 } //______________________________________________
 void MatchTOF::addTRDSeed(const o2::trd::TrackTRD& _tr, o2::dataformats::GlobalTrackID srcGID, float time0, float terr)
@@ -471,7 +617,6 @@ void MatchTOF::addTRDSeed(const o2::trd::TrackTRD& _tr, o2::dataformats::GlobalT
   auto trc = _tr.getOuterParam();
 
   o2::track::TrackLTIntegral intLT0 = _tr.getLTIntegralOut();
-  ; // empty for the moment
 
   // o2::dataformats::TimeStampWithError<float, float>
   timeEst ts(time0, terr + mExtraTimeToleranceTRD);
@@ -485,9 +630,6 @@ void MatchTOF::addTPCSeed(const o2::tpc::TrackTPC& _tr, o2::dataformats::GlobalT
 
   std::array<float, 3> globalPos;
 
-  // current track index
-  int it = mTracksWork[trkType::UNCONS].size();
-
   // create working copy of track param
   timeEst timeInfo;
   // set
@@ -496,65 +638,66 @@ void MatchTOF::addTPCSeed(const o2::tpc::TrackTPC& _tr, o2::dataformats::GlobalT
     extraErr = 100;
   }
 
+  float trackTime0 = _tr.getTime0() * mTPCTBinMUS;
+  timeInfo.setTimeStampError((_tr.getDeltaTBwd() + 5) * mTPCTBinMUS + extraErr);
+  //  timeInfo.setTimeStampError(trackTime0 - time0 + terr + extraErr);
+  timeInfo.setTimeStamp(trackTime0);
+
   auto trc = _tr.getOuterParam();
-  const auto& trackTune = o2::globaltracking::TrackTuneParams::Instance();
-  if (!trackTune.sourceLevelTPC) { // correct only if TPC track was not corrected at the source level
-    if (trackTune.useTPCOuterCorr) {
-      trc.updateParams(trackTune.tpcParOuter);
-    }
-    if (trackTune.tpcCovOuterType != o2::globaltracking::TrackTuneParams::AddCovType::Disable) { // only TRD-refitted track have cov.matrix already manipulated
-      trc.updateCov(trackTune.tpcCovOuter, trackTune.tpcCovOuterType == o2::globaltracking::TrackTuneParams::AddCovType::WithCorrelations);
-    }
-  }
-  if (!propagateToRefXWithoutCov(trc, mXRef, 10, mBz)) { // we first propagate to 371 cm without considering the covariance matri
-    mNotPropagatedToTOF[trkType::UNCONS]++;
-    return;
-  }
+  trc.getXYZGlo(globalPos);
 
-  o2::track::TrackLTIntegral intLT0; //mTPCTracksWork.back().getLTIntegralOut(); // we get the integrated length from TPC-ITC outward propagation
+  int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
+
+  mSideTPC[sector].push_back(_tr.hasASideClustersOnly() ? 1 : (_tr.hasCSideClustersOnly() ? -1 : 0));
+  mExtraTPCFwdTime[sector].push_back((_tr.getDeltaTFwd() + 5) * mTPCTBinMUS + extraErr);
+  //  mExtraTPCFwdTime[sector].push_back(time0 + terr - trackTime0 + extraErr);
+  mTracksWork[sector][trkType::UNCONS].emplace_back(std::make_pair(trc, timeInfo));
+  mTrackGid[sector][trkType::UNCONS].emplace_back(srcGID);
+
+  if (mMCTruthON) {
+    mTracksLblWork[sector][trkType::UNCONS].emplace_back(mRecoCont->getTPCTrackMCLabel(srcGID));
+  }
+  o2::track::TrackLTIntegral intLT0; // mTPCTracksWork.back().getLTIntegralOut(); // we get the integrated length from TPC-ITC outward propagation
   // compute track length up to now
-  o2::base::Propagator::Instance()->estimateLTFast(intLT0, trc);
+  mLTinfos[sector][trkType::UNCONS].emplace_back(intLT0);
 
-  if (trc.getX() < o2::constants::geom::XTPCOuterRef - 1.) {
-    if (!propagateToRefX(trc, o2::constants::geom::XTPCOuterRef, 10, intLT0) || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix worked; CHECK: can it happ
+  /*
+    const auto& trackTune = o2::globaltracking::TrackTuneParams::Instance();
+    if (!trackTune.sourceLevelTPC) { // correct only if TPC track was not corrected at the source level
+      if (trackTune.useTPCOuterCorr) {
+        trc.updateParams(trackTune.tpcParOuter);
+      }
+      if (trackTune.tpcCovOuterType != o2::globaltracking::TrackTuneParams::AddCovType::Disable) { // only TRD-refitted track have cov.matrix already manipulated
+        trc.updateCov(trackTune.tpcCovOuter, trackTune.tpcCovOuterType == o2::globaltracking::TrackTuneParams::AddCovType::WithCorrelations);
+      }
+    }
+    if (!propagateToRefXWithoutCov(trc, mXRef, 10, mBz)) { // we first propagate to 371 cm without considering the covariance matri
       mNotPropagatedToTOF[trkType::UNCONS]++;
       return;
     }
-  }
 
-  // the "rough" propagation worked; now we can propagate considering also the cov matrix
-  if (!propagateToRefX(trc, mXRef, 2, intLT0)) { // || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix worked; CHECK: can it happen that it does not if the prop>
-    mNotPropagatedToTOF[trkType::UNCONS]++;
-    return;
-  }
+    if (trc.getX() < o2::constants::geom::XTPCOuterRef - 1.) {
+      if (!propagateToRefX(trc, o2::constants::geom::XTPCOuterRef, 10, intLT0) || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix worked; CHECK: can it happ
+        mNotPropagatedToTOF[trkType::UNCONS]++;
+        return;
+      }
+    }
 
-  float trackTime0 = _tr.getTime0() * mTPCTBinMUS;
+    // the "rough" propagation worked; now we can propagate considering also the cov matrix
+    if (!propagateToRefX(trc, mXRef, 2, intLT0)) { // || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) { // we check that the propagation with the cov matrix worked; CHECK: can it happen that it does not if the prop>
+      mNotPropagatedToTOF[trkType::UNCONS]++;
+      return;
+    }
 
-  timeInfo.setTimeStampError((_tr.getDeltaTBwd() + 5) * mTPCTBinMUS + extraErr);
-  mExtraTPCFwdTime.push_back((_tr.getDeltaTFwd() + 5) * mTPCTBinMUS + extraErr);
+    //  printf("time0 %f -> %f (diff = %f, err = %f)\n",trackTime0, time0, trackTime0 - time0, terr);
+    //  printf("time errors %f,%f -> %f,%f\n",(_tr.getDeltaTBwd() + 5) * mTPCTBinMUS,(_tr.getDeltaTFwd() + 5) * mTPCTBinMUS,trackTime0 - time0 + terr, time0 + terr - trackTime0);
 
-  //  timeInfo.setTimeStampError(trackTime0 - time0 + terr + extraErr);
-  //  mExtraTPCFwdTime.push_back(time0 + terr - trackTime0 + extraErr);
+    trc.getXYZGlo(globalPos);
+    int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
 
-  timeInfo.setTimeStamp(trackTime0);
-  mSideTPC.push_back(_tr.hasASideClustersOnly() ? 1 : (_tr.hasCSideClustersOnly() ? -1 : 0));
-
-  //  printf("time0 %f -> %f (diff = %f, err = %f)\n",trackTime0, time0, trackTime0 - time0, terr);
-  //  printf("time errors %f,%f -> %f,%f\n",(_tr.getDeltaTBwd() + 5) * mTPCTBinMUS,(_tr.getDeltaTFwd() + 5) * mTPCTBinMUS,trackTime0 - time0 + terr, time0 + terr - trackTime0);
-
-  trc.getXYZGlo(globalPos);
-  int sector = o2::math_utils::angle2Sector(TMath::ATan2(globalPos[1], globalPos[0]));
-
-  mTracksWork[trkType::UNCONS].emplace_back(std::make_pair(trc, timeInfo));
-  mTrackGid[trkType::UNCONS].emplace_back(srcGID);
-
-  if (mMCTruthON) {
-    mTracksLblWork[trkType::UNCONS].emplace_back(mRecoCont->getTPCTrackMCLabel(srcGID));
-  }
-  mLTinfos[trkType::UNCONS].emplace_back(intLT0);
-
-  mTracksSectIndexCache[trkType::UNCONS][sector].push_back(it);
-  //delete trc; // Check: is this needed?
+    mTracksSectIndexCache[trkType::UNCONS][sector].push_back(it);
+  */
+  // delete trc; // Check: is this needed?
 }
 //______________________________________________
 bool MatchTOF::prepareTOFClusters()
@@ -573,7 +716,7 @@ bool MatchTOF::prepareTOFClusters()
   //    mTOFClusLblWork.reserve(mNumOfClusters);
   //  }
 
-  for (int sec = o2::constants::math::NSectors; sec--;) {
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
     mTOFClusSectIndexCache[sec].clear();
     //mTOFClusSectIndexCache[sec].reserve(100 + 1.2 * mNumOfClusters / o2::constants::math::NSectors); // we cannot do this, we don't have mNumOfClusters yet
   }
@@ -594,7 +737,7 @@ bool MatchTOF::prepareTOFClusters()
   }
 
   // sort clusters in each sector according to their time (increasing in time)
-  for (int sec = o2::constants::math::NSectors; sec--;) {
+  for (int sec = o2::constants::math::NSectors - 1; sec > -1; sec--) {
     auto& indexCache = mTOFClusSectIndexCache[sec];
     LOG(debug) << "Sorting sector" << sec << " | " << indexCache.size() << " TOF clusters";
     if (!indexCache.size()) {
@@ -647,9 +790,9 @@ void MatchTOF::doMatching(int sec)
       nStepsInsideSameStrip[ii] = 0;
     }
     int nStripsCrossedInPropagation = 0; // how many strips were hit during the propagation
-    auto& trackWork = mTracksWork[type][cacheTrk[itrk]];
+    auto& trackWork = mTracksWork[sec][type][cacheTrk[itrk]];
     auto& trefTrk = trackWork.first;
-    auto& intLT = mLTinfos[type][cacheTrk[itrk]];
+    auto& intLT = mLTinfos[sec][type][cacheTrk[itrk]];
     float timeShift = intLT.getL() * 33.35641; // integrated time for 0.75 beta particles in ps, to take into account the t.o.f. delay with respect the interaction BC
                                                // using beta=0.75 to cover beta range [0.59 , 1.04] also for a 8 m track lenght with a 10 ns track resolution (TRD)
 
@@ -850,7 +993,7 @@ void MatchTOF::doMatching(int sec)
           foundCluster = true;
           // set event indexes (to be checked)
           int eventIndexTOFCluster = mTOFClusSectIndexCache[indices[0]][itof];
-          mMatchedTracksPairs.emplace_back(cacheTrk[itrk], eventIndexTOFCluster, mTOFClusWork[cacheTOF[itof]].getTime(), chi2, trkLTInt[iPropagation], mTrackGid[type][cacheTrk[itrk]], type, (trefTOF.getTime() - (minTrkTime + maxTrkTime - 100E3) * 0.5) * 1E-6, 0., resX, resZ); // subracting 100 ns to max track which was artificially added
+          mMatchedTracksPairsSec[sec].emplace_back(cacheTrk[itrk], eventIndexTOFCluster, mTOFClusWork[cacheTOF[itof]].getTime(), chi2, trkLTInt[iPropagation], mTrackGid[sec][type][cacheTrk[itrk]], type, (trefTOF.getTime() - (minTrkTime + maxTrkTime - 100E3) * 0.5) * 1E-6, 0., resX, resZ); // subracting 100 ns to max track which was artificially added
         }
       }
     }
@@ -894,9 +1037,9 @@ void MatchTOF::doMatchingForTPC(int sec)
   LOG(debug) << "Trying to match %d tracks" << cacheTrk.size();
 
   for (int itrk = 0; itrk < cacheTrk.size(); itrk++) {
-    auto& trackWork = mTracksWork[trkType::UNCONS][cacheTrk[itrk]];
+    auto& trackWork = mTracksWork[sec][trkType::UNCONS][cacheTrk[itrk]];
     auto& trefTrk = trackWork.first;
-    auto& intLT = mLTinfos[trkType::UNCONS][cacheTrk[itrk]];
+    auto& intLT = mLTinfos[sec][trkType::UNCONS][cacheTrk[itrk]];
 
     float timeShift = intLT.getL() * 33.35641; // integrated time for 0.75 beta particles in ps, to take into account the t.o.f. delay with respect the interaction BC
                                                // using beta=0.75 to cover beta range [0.59 , 1.04] also for a 8 m track lenght with a 10 ns track resolution (TRD)
@@ -904,11 +1047,11 @@ void MatchTOF::doMatchingForTPC(int sec)
     BCcand.clear();
     nStripsCrossedInPropagation.clear();
 
-    int side = mSideTPC[cacheTrk[itrk]];
+    int side = mSideTPC[sec][cacheTrk[itrk]];
     // look at BC candidates for the track
     double minTrkTime = (trackWork.second.getTimeStamp() - trackWork.second.getTimeStampError()) * 1.E6 + timeShift; // minimum time in ps
     minTrkTime = int(minTrkTime / BCgranularity) * BCgranularity;                                                    // align min to a BC
-    double maxTrkTime = (trackWork.second.getTimeStamp() + mExtraTPCFwdTime[cacheTrk[itrk]]) * 1.E6 + timeShift;     // maximum time in ps
+    double maxTrkTime = (trackWork.second.getTimeStamp() + mExtraTPCFwdTime[sec][cacheTrk[itrk]]) * 1.E6 + timeShift; // maximum time in ps
 
     if (mIsCosmics) {
       for (double tBC = minTrkTime; tBC < maxTrkTime; tBC += BCgranularity) {
@@ -1179,7 +1322,7 @@ void MatchTOF::doMatchingForTPC(int sec)
             foundCluster = true;
             // set event indexes (to be checked)
             int eventIndexTOFCluster = mTOFClusSectIndexCache[indices[0]][itof];
-            mMatchedTracksPairs.emplace_back(cacheTrk[itrk], eventIndexTOFCluster, mTOFClusWork[cacheTOF[itof]].getTime(), chi2, trkLTInt[ibc][iPropagation], mTrackGid[trkType::UNCONS][cacheTrk[itrk]], trkType::UNCONS, resZ / mTPCVDrift * side, trefTOF.getZ(), resX, resZ); // TODO: check if this is correct!
+            mMatchedTracksPairsSec[sec].emplace_back(cacheTrk[itrk], eventIndexTOFCluster, mTOFClusWork[cacheTOF[itof]].getTime(), chi2, trkLTInt[ibc][iPropagation], mTrackGid[sec][trkType::UNCONS][cacheTrk[itrk]], trkType::UNCONS, resZ / mTPCVDrift * side, trefTOF.getZ(), resX, resZ); // TODO: check if this is correct!
           }
         }
       }
@@ -1231,13 +1374,13 @@ int MatchTOF::findFITIndex(int bc, const gsl::span<const o2::ft0::RecPoints>& FI
   return index;
 }
 //______________________________________________
-void MatchTOF::selectBestMatches()
+void MatchTOF::selectBestMatches(int sec)
 {
   if (mSetHighPurity) {
-    BestMatchesHP(mMatchedTracksPairs, mMatchedTracks, mMatchedTracksIndex, mMatchedClustersIndex, mFITRecPoints, mTOFClusWork, mCalibInfoTOF, mTimestamp, mMCTruthON, mTOFClusLabels, mTracksLblWork, mOutTOFLabels);
+    BestMatchesHP(mMatchedTracksPairsSec[sec], mMatchedTracks, mMatchedTracksIndex[sec], mMatchedClustersIndex, mFITRecPoints, mTOFClusWork, mCalibInfoTOF, mTimestamp, mMCTruthON, mTOFClusLabels, mTracksLblWork[sec], mOutTOFLabels);
     return;
   }
-  BestMatches(mMatchedTracksPairs, mMatchedTracks, mMatchedTracksIndex, mMatchedClustersIndex, mFITRecPoints, mTOFClusWork, mTracksWork, mCalibInfoTOF, mTimestamp, mMCTruthON, mTOFClusLabels, mTracksLblWork, mOutTOFLabels, mMatchParams->calibMaxChi2);
+  BestMatches(mMatchedTracksPairsSec[sec], mMatchedTracks, mMatchedTracksIndex[sec], mMatchedClustersIndex, mFITRecPoints, mTOFClusWork, mTracksWork[sec], mCalibInfoTOF, mTimestamp, mMCTruthON, mTOFClusLabels, mTracksLblWork[sec], mOutTOFLabels, mMatchParams->calibMaxChi2);
 }
 //______________________________________________
 void MatchTOF::BestMatches(std::vector<o2::dataformats::MatchInfoTOFReco>& matchedTracksPairs, std::vector<o2::dataformats::MatchInfoTOF>* matchedTracks, std::vector<int>* matchedTracksIndex, int* matchedClustersIndex, const gsl::span<const o2::ft0::RecPoints>& FITRecPoints, const std::vector<Cluster>& TOFClusWork, const std::vector<matchTrack>* TracksWork, std::vector<o2::dataformats::CalibInfoTOF>& CalibInfoTOF, unsigned long Timestamp, bool MCTruthON, const o2::dataformats::MCTruthContainer<o2::MCCompLabel>* TOFClusLabels, const std::vector<o2::MCCompLabel>* TracksLblWork, std::vector<o2::MCCompLabel>* OutTOFLabels, float calibMaxChi2)

--- a/Detectors/GlobalTrackingWorkflow/include/GlobalTrackingWorkflow/TOFMatcherSpec.h
+++ b/Detectors/GlobalTrackingWorkflow/include/GlobalTrackingWorkflow/TOFMatcherSpec.h
@@ -29,7 +29,7 @@ namespace globaltracking
 {
 
 /// create a processor spec
-framework::DataProcessorSpec getTOFMatcherSpec(o2::dataformats::GlobalTrackID::mask_t src, bool useMC, bool useFIT, bool tpcRefit, bool strict, float extratolerancetrd, bool pushMatchable, const o2::tpc::CorrectionMapsLoaderGloOpts& sclOpts);
+framework::DataProcessorSpec getTOFMatcherSpec(o2::dataformats::GlobalTrackID::mask_t src, bool useMC, bool useFIT, bool tpcRefit, bool strict, float extratolerancetrd, bool pushMatchable, const o2::tpc::CorrectionMapsLoaderGloOpts& sclOpts, int nlanes = 1);
 
 } // namespace globaltracking
 } // namespace o2

--- a/Detectors/GlobalTrackingWorkflow/src/TOFMatcherSpec.cxx
+++ b/Detectors/GlobalTrackingWorkflow/src/TOFMatcherSpec.cxx
@@ -58,7 +58,7 @@ namespace globaltracking
 class TOFMatcherSpec : public Task
 {
  public:
-  TOFMatcherSpec(std::shared_ptr<DataRequest> dr, std::shared_ptr<o2::base::GRPGeomRequest> gr, bool useMC, bool useFIT, bool tpcRefit, bool strict, bool pushMatchable) : mDataRequest(dr), mGGCCDBRequest(gr), mUseMC(useMC), mUseFIT(useFIT), mDoTPCRefit(tpcRefit), mStrict(strict), mPushMatchable(pushMatchable) {}
+  TOFMatcherSpec(std::shared_ptr<DataRequest> dr, std::shared_ptr<o2::base::GRPGeomRequest> gr, bool useMC, bool useFIT, bool tpcRefit, bool strict, bool pushMatchable, int lanes = 1) : mDataRequest(dr), mGGCCDBRequest(gr), mUseMC(useMC), mUseFIT(useFIT), mDoTPCRefit(tpcRefit), mStrict(strict), mPushMatchable(pushMatchable), mNlanes(lanes) {}
   ~TOFMatcherSpec() override = default;
   void init(InitContext& ic) final;
   void run(ProcessingContext& pc) final;
@@ -77,6 +77,7 @@ class TOFMatcherSpec : public Task
   bool mStrict = false;
   bool mPushMatchable = false;
   float mExtraTolTRD = 0.;
+  int mNlanes = 1;
   MatchTOF mMatcher; ///< Cluster finder
   TStopwatch mTimer;
 };
@@ -92,6 +93,7 @@ void TOFMatcherSpec::init(InitContext& ic)
   mTPCCorrMapsLoader.init(ic);
   mMatcher.storeMatchable(mPushMatchable);
   mMatcher.setExtraTimeToleranceTRD(mExtraTolTRD);
+  mMatcher.setNlanes(mNlanes);
 }
 
 void TOFMatcherSpec::updateTimeDependentParams(ProcessingContext& pc)
@@ -237,7 +239,7 @@ void TOFMatcherSpec::endOfStream(EndOfStreamContext& ec)
        mTimer.CpuTime(), mTimer.RealTime(), mTimer.Counter() - 1);
 }
 
-DataProcessorSpec getTOFMatcherSpec(GID::mask_t src, bool useMC, bool useFIT, bool tpcRefit, bool strict, float extratolerancetrd, bool pushMatchable, const o2::tpc::CorrectionMapsLoaderGloOpts& sclOpts)
+DataProcessorSpec getTOFMatcherSpec(GID::mask_t src, bool useMC, bool useFIT, bool tpcRefit, bool strict, float extratolerancetrd, bool pushMatchable, const o2::tpc::CorrectionMapsLoaderGloOpts& sclOpts, int nlanes)
 {
   uint32_t ss = o2::globaltracking::getSubSpec(strict ? o2::globaltracking::MatchingType::Strict : o2::globaltracking::MatchingType::Standard);
   Options opts;
@@ -314,7 +316,7 @@ DataProcessorSpec getTOFMatcherSpec(GID::mask_t src, bool useMC, bool useFIT, bo
     "tof-matcher",
     dataRequest->inputs,
     outputs,
-    AlgorithmSpec{adaptFromTask<TOFMatcherSpec>(dataRequest, ggRequest, useMC, useFIT, tpcRefit, strict, pushMatchable)},
+    AlgorithmSpec{adaptFromTask<TOFMatcherSpec>(dataRequest, ggRequest, useMC, useFIT, tpcRefit, strict, pushMatchable, nlanes)},
     opts};
 }
 

--- a/Detectors/GlobalTrackingWorkflow/src/tof-matcher-workflow.cxx
+++ b/Detectors/GlobalTrackingWorkflow/src/tof-matcher-workflow.cxx
@@ -55,6 +55,7 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
   // option allowing to set parameters
   std::vector<o2::framework::ConfigParamSpec> options{
     {"disable-mc", o2::framework::VariantType::Bool, false, {"disable MC propagation even if available"}},
+    {"tof-lanes", o2::framework::VariantType::Int, 1, {"number of parallel lanes up to the matcher"}},
     {"disable-root-input", o2::framework::VariantType::Bool, false, {"disable root-files input reader"}},
     {"disable-root-output", o2::framework::VariantType::Bool, false, {"disable root-files output writer"}},
     {"track-sources", VariantType::String, std::string{GID::ALL}, {"comma-separated list of sources to use: allowed TPC,ITS-TPC,TPC-TRD,ITS-TPC-TRD (all)"}},
@@ -98,6 +99,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
   bool writematching = 0;
   bool writecalib = 0;
   auto outputType = configcontext.options().get<std::string>("output-type");
+  auto nLanes = configcontext.options().get<int>("tof-lanes");
   if (outputType.rfind("matching-info") < outputType.size()) {
     writematching = 1;
   }
@@ -124,6 +126,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
   LOG(debug) << "TOF matching in strict mode = " << strict;
   LOG(debug) << "TOF extra time tolerance for TRD tracks = " << extratolerancetrd;
   LOG(debug) << "Store all matchables = " << writeMatchable;
+  LOG(debug) << "TOF Nlanes for matcher = " << nLanes;
 
   //GID::mask_t alowedSources = GID::getSourcesMask("TPC,ITS-TPC");
   GID::mask_t alowedSources = GID::getSourcesMask("TPC,ITS-TPC,TPC-TRD,ITS-TPC-TRD");
@@ -160,7 +163,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
     }
   }
 
-  specs.emplace_back(o2::globaltracking::getTOFMatcherSpec(src, useMC, useFIT, false, strict, extratolerancetrd, writeMatchable, sclOpt)); // doTPCrefit not yet supported (need to load TPC clusters?)
+  specs.emplace_back(o2::globaltracking::getTOFMatcherSpec(src, useMC, useFIT, false, strict, extratolerancetrd, writeMatchable, sclOpt, nLanes)); // doTPCrefit not yet supported (need to load TPC clusters?)
 
   if (!disableRootOut) {
     std::vector<DataProcessorSpec> writers;

